### ISSUE TITLE
Docs: modify examples for explicit-module-boundary-types

### ISSUE THREAD

**Issue by Ryan-Dia** [typescript-eslint/typescript-eslint#7430]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

1.
![image](https://github.com/typescript-eslint/typescript-eslint/assets/76567238/febd1f49-1ac5-4718-9a34-a09fc54c9ede)

2.
![image](https://github.com/typescript-eslint/typescript-eslint/assets/76567238/a5f15f99-8ce9-4c6d-89b6-7d6945310e8c)

3.
![image](https://github.com/typescript-eslint/typescript-eslint/assets/76567238/6cedb342-dabf-4761-a59d-77d6f94b0dc7)

I had difficulty understanding the `explicit-module-boundary-types` rule when I first encountered it due to unclear and non-intuitive examples.

In the initial description of this rule, it states, "Require explicit return and argument types on exported functions and classes public class methods." When writing code that violates this rule in the playground, it shows an error message like "Missing return type on function," as seen in the image above. Furthermore, in the third image, the comment for Incorrect code reads, "Should indicate that no value is returned (void)."

Based on this information, one way to address this issue is to avoid exporting meaningless functions or classes. However, to better understand and adhere to this rule in a clear and intuitive manner, I believe it is more appropriate to follow the approach described in the document: explicitly specifying types to avoid errors. By doing so, we can easily grasp the essence of the rule, which emphasizes "Requiring explicit return and argument types on exported functions and classes public class methods", creating more straightforward and intuitive examples.

# before

### ✅ Correct

```ts
// Function is not exported
function test() {
  return;
}

// A return value of type number
export var fn = function (): number {
  return 1;
};
// A return value of type string
export var arrowFn = (): string => 'test';
// All arguments should be typed
export var arrowFn = (arg: string): string => `test ${arg}`;
export var arrowFn = (arg: unknown): string => `test ${arg}`;

// Class is not exported
class Test {
  method() {
    return;
  }
}
```

# after

### ✅ Correct

```ts
// A function with no return value (void)
export function test(): void {
  return;
}

// A return value of type number
export var fn = function (): number {
  return 1;
};
// A return value of type string
export var arrowFn = (): string => 'test';
// All arguments should be typed
export var arrowFn = (arg: string): string => `test ${arg}`;
export var arrowFn = (arg: unknown): string => `test ${arg}`;

// A class method with no return value (void)
export class Test {
  method(): void {
    return;
  }
}
```

## 📌 Answer : Josh-Cena

<img width="926" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/76567238/3b2255dd-8eb3-4e3e-8b8a-c1987865d029">

## 📌 Answer : ryan-dia

When I first encountered these two points in the rule, having both of them in a single example made it harder to understand. Therefore, as you suggested, instead of putting both points in one example, dividing them into separate examples would make it much more intuitive, clear, and easier to comprehend.   [applied code](https://github.com/typescript-eslint/typescript-eslint/pull/7404/commits/8eaf4cd07005d20195d70574c5d7933ed56ffb14)

Alternatively, in my opinion, since the rule name itself is not "explicit-types" but "explicit-module-boundary-types," it is natural to assume that it won't apply unless it's related to module boundaries. Hence, there might be no need to include examples for cases outside module boundaries. Moreover, in the initial explanation of the rule, it states "on exported functions' and classes' public class methods," and in the "When Not To Use It" section, it mentions, "If you wish to make sure all functions have explicit return types, as opposed to only the module boundaries, you can use explicit-function-return-type." This clearly indicates that the rule applies only to module boundaries. Users can infer this, and if you still wish to include the examples, it would be better to present them as follows.

# 🛠️  case 1

### ❌ Incorrect

```ts
// Should indicate that no value is returned (void)
export function test() {
  return;
}

// Should indicate that a number is returned
export default function () {
  return 1;
}

// Should indicate that a string is returned
export var arrowFn = () => 'test';

// All arguments should be typed
export var arrowFn = (arg): string => `test ${arg}`;
export var arrowFn = (arg: any): string => `test ${arg}`;

export class Test {
  // Should indicate that no value is returned (void)
  method() {
    return;
  }
}

// The function does not apply because it is not an exported function.
function test() {
  return;
}
```

### ✅ Correct

```ts
// A function with no return value (void)
export function test(): void {
  return;
}

// A return value of type number
export var fn = function (): number {
  return 1;
};
// A return value of type string
export var arrowFn = (): string => 'test';
// All arguments should be typed
export var arrowFn = (arg: string): string => `test ${arg}`;
export var arrowFn = (arg: unknown): string => `test ${arg}`;

// A class method with no return value (void)
export class Test {
  method(): void {
    return;
  }
}

// The function does not apply because it is not an exported function.
function test() {
  return;
}
```

# 🛠️ case 2

### ❌ Incorrect

```ts
// Should indicate that no value is returned (void)
export function test() {
  return;
}

// Should indicate that a number is returned
export default function () {
  return 1;
}

// Should indicate that a string is returned
export var arrowFn = () => 'test';

// All arguments should be typed
export var arrowFn = (arg): string => `test ${arg}`;
export var arrowFn = (arg: any): string => `test ${arg}`;

export class Test {
  // Should indicate that no value is returned (void)
  method() {
    return;
  }
}
```

### ✅ Correct

```ts
// If you remove the export, you have the option to either write explicit return and argument types or not. 

// A function with no return value (void)
export function test(): void {
  return;
}

// A return value of type number
export var fn = function (): number {
  return 1;
};
// A return value of type string
export var arrowFn = (): string => 'test';
// All arguments should be typed
export var arrowFn = (arg: string): string => `test ${arg}`;
export var arrowFn = (arg: unknown): string => `test ${arg}`;

// A class method with no return value (void)
export class Test {
  method(): void {
    return;
  }
}
```

### Affected URL(s)

https://typescript-eslint.io/rules/explicit-module-boundary-types/

### Reference
#7404

**Comment by JoshuaKGoldberg** [comment]
Note: please don't post screencap pictures. They don't show up in search results and are not accessible to folks who can't visually perceive them. 

(for reference, the screenshots are of the existing rule docs in production now)

**Comment by JoshuaKGoldberg** [comment]
Surfacing @Josh-Cena's https://github.com/typescript-eslint/typescript-eslint/pull/7404#pullrequestreview-1561089210 here:

> The point of this rule is:
> 
> 1. Exported members must have explicit signatures
> 2. Non-exported members may or may not have explicit signatures
> 
> Your change makes everything exported, so the second case is lost. I'm not sure how this is an improvement. If you wish, though, we can have both an exported and a non-exported example.

Agreed. It's good to have symmetry in incorrect/correct examples so readers can understand how to fix incorrect code.

**Comment by Ryan-Dia** [comment]
> Note: please don't post screencap pictures. They don't show up in search results and are not accessible to folks who can't visually perceive them.
> 
> (for reference, the screenshots are of the existing rule docs in production now)

Thank you for letting me know. I hadn't considered that aspect; I apologize."
